File: discord-bot-cli/discord_bot_cli/bot.py
"""
Discord bot implementation
"""

import asyncio
import logging
import sys
from typing import Optional, Callable, Any
import discord
from discord.ext import commands

from .config import Config


class DiscordBot:
    """Main Discord bot class"""
    
    def __init__(self, config: Config):
        self.config = config
        self.bot = commands.Bot(
            command_prefix=config.bot.prefix,
            intents=discord.Intents.default()
        )
        self.bot.intents.message_content = True
        
        # Message handlers
        self.message_handlers: list[Callable[[discord.Message], Any]] = []
        self.ready_handlers: list[Callable[[], Any]] = []
        
        # Setup event handlers
        self._setup_handlers()
        
        # Reconnection state
        self.reconnect_attempts = 0
        self.is_running = False
    
    def _setup_handlers(self):
        """Setup Discord event handlers"""
        
        @self.bot.event
        async def on_ready():
            """Called when bot is ready"""
            logging.info(f"Bot logged in as {self.bot.user}")
            self.reconnect_attempts = 0
            self.is_running = True
            
            # Call registered ready handlers
            for handler in self.ready_handlers:
                try:
                    if asyncio.iscoroutinefunction(handler):
                        await handler()
                    else:
                        handler()
                except Exception as e:
                    logging.error(f"Error in ready handler: {e}")
        
        @self.bot.event
        async def on_message(message):
            """Handle incoming messages"""
            # Ignore messages from the bot itself
            if message.author == self.bot.user:
                return
            
            # Check if message is in the configured channel
            if self.config.bot.channel_id and message.channel.id != self.config.bot.channel_id:
                return
            
            # Call registered message handlers
            for handler in self.message_handlers:
                try:
                    if asyncio.iscoroutinefunction(handler):
                        await handler(message)
                    else:
                        handler(message)
                except Exception as e:
                    logging.error(f"Error in message handler: {e}")
            
            # Process commands
            await self.bot.process_commands(message)
        
        @self.bot.event
        async def on_disconnect():
            """Called when bot disconnects"""
            logging.warning("Bot disconnected")
            self.is_running = False
        
        @self.bot.event
        async def on_error(event, *args, **kwargs):
            """Handle errors"""
            logging.error(f"Error in event {event}: {sys.exc_info()}")
    
    def add_message_handler(self, handler: Callable[[discord.Message], Any]):
        """Add a message handler"""
        self.message_handlers.append(handler)
    
    def add_ready_handler(self, handler: Callable[[], Any]):
        """Add a ready handler"""
        self.ready_handlers.append(handler)
    
    async def start(self):
        """Start the bot"""
        try:
            await self.bot.start(self.config.bot.token)
        except discord.LoginFailure:
            logging.error("Invalid Discord bot token")
            raise
        except Exception as e:
            logging.error(f"Error starting bot: {e}")
            raise
    
    async def stop(self):
        """Stop the bot"""
        self.is_running = False
        await self.bot.close()
    
    async def send_message(self, content: str, channel_id: Optional[int] = None) -> bool:
        """Send a message to Discord"""
        try:
            channel_id = channel_id or self.config.bot.channel_id
            if not channel_id:
                logging.error("No channel ID configured")
                return False
            
            channel = self.bot.get_channel(channel_id)
            if not channel:
                logging.error(f"Channel {channel_id} not found")
                return False
            
            await channel.send(content)
            return True
        except Exception as e:
            logging.error(f"Error sending message: {e}")
            return False
    
    async def get_channel_messages(self, limit: int = 10) -> list[discord.Message]:
        """Get recent messages from the configured channel"""
        try:
            channel_id = self.config.bot.channel_id
            if not channel_id:
                logging.error("No channel ID configured")
                return []
            
            channel = self.bot.get_channel(channel_id)
            if not channel:
                logging.error(f"Channel {channel_id} not found")
                return []
            
            messages = []
            async for message in channel.history(limit=limit):
                messages.append(message)
            
            return messages
        except Exception as e:
            logging.error(f"Error getting channel messages: {e}")
            return []
    
    def is_connected(self) -> bool:
        """Check if bot is connected to Discord"""
        return self.bot.is_ready() and self.is_running

File: discord-bot-cli-js/src/bot.js
/**
 * Discord bot implementation
 */

const { Client, GatewayIntentBits, Events } = require('discord.js');
const EventEmitter = require('events');

class DiscordBot extends EventEmitter {
  constructor(config, logger) {
    super();
    this.config = config;
    this.logger = logger;
    this.client = null;
    this.isConnected = false;
    this.reconnectAttempts = 0;
    this.messageHandlers = [];
    this.readyHandlers = [];
  }

  async initialize() {
    this.client = new Client({
      intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildMessages,
        GatewayIntentBits.MessageContent,
        GatewayIntentBits.GuildMembers
      ]
    });

    this.setupEventHandlers();
  }

  setupEventHandlers() {
    this.client.once(Events.ClientReady, (readyClient) => {
      this.logger.info(`Bot logged in as ${readyClient.user.tag}`);
      this.isConnected = true;
      this.reconnectAttempts = 0;
      this.emit('ready');
      
      // Call registered ready handlers
      this.readyHandlers.forEach(handler => {
        try {
          if (typeof handler === 'function') {
            handler();
          }
        } catch (error) {
          this.logger.error('Error in ready handler:', error);
        }
      });
    });

    this.client.on(Events.MessageCreate, (message) => {
      // Ignore messages from the bot itself
      if (message.author.bot) return;

      // Check if message is in the configured channel
      if (this.config.bot.channelId && message.channel.id !== this.config.bot.channelId) {
        return;
      }

      this.logger.debug(`Received message from ${message.author.tag}: ${message.content}`);

      // Call registered message handlers
      this.messageHandlers.forEach(handler => {
        try {
          if (typeof handler === 'function') {
            handler(message);
          }
        } catch (error) {
          this.logger.error('Error in message handler:', error);
        }
      });

      this.emit('message', message);
    });

    this.client.on(Events.Error, (error) => {
      this.logger.error('Discord client error:', error);
      this.emit('error', error);
    });

    this.client.on(Events.Disconnect, () => {
      this.logger.warn('Bot disconnected');
      this.isConnected = false;
      this.emit('disconnect');
    });

    this.client.on(Events.ShardError, (error) => {
      this.logger.error('Shard error:', error);
      this.emit('error', error);
    });
  }

  async start() {
    try {
      if (!this.client) {
        await this.initialize();
      }

      await this.client.login(this.config.bot.token);
    } catch (error) {
      this.logger.error('Failed to start bot:', error);
      throw error;
    }
  }

  async stop() {
    if (this.client) {
      this.isConnected = false;
      await this.client.destroy();
      this.client = null;
    }
  }

  async sendMessage(content, channelId = null) {
    try {
      const targetChannelId = channelId || this.config.bot.channelId;
      if (!targetChannelId) {
        this.logger.error('No channel ID configured');
        return false;
      }

      const channel = this.client.channels.cache.get(targetChannelId);
      if (!channel) {
        this.logger.error(`Channel ${targetChannelId} not found`);
        return false;
      }

      await channel.send(content);
      this.logger.debug(`Sent message to channel ${targetChannelId}: ${content}`);
      return true;
    } catch (error) {
      this.logger.error('Error sending message:', error);
      return false;
    }
  }

  async getChannelMessages(limit = 10) {
    try {
      const channelId = this.config.bot.channelId;
      if (!channelId) {
        this.logger.error('No channel ID configured');
        return [];
      }

      const channel = this.client.channels.cache.get(channelId);
      if (!channel) {
        this.logger.error(`Channel ${channelId} not found`);
        return [];
      }

      const messages = await channel.messages.fetch({ limit });
      return Array.from(messages.values());
    } catch (error) {
      this.logger.error('Error getting channel messages:', error);
      return [];
    }
  }

  addMessageHandler(handler) {
    if (typeof handler === 'function') {
      this.messageHandlers.push(handler);
    }
  }

  addReadyHandler(handler) {
    if (typeof handler === 'function') {
      this.readyHandlers.push(handler);
    }
  }

  removeMessageHandler(handler) {
    const index = this.messageHandlers.indexOf(handler);
    if (index > -1) {
      this.messageHandlers.splice(index, 1);
    }
  }

  removeReadyHandler(handler) {
    const index = this.readyHandlers.indexOf(handler);
    if (index > -1) {
      this.readyHandlers.splice(index, 1);
    }
  }

  getConnectionStatus() {
    return {
      connected: this.isConnected,
      ready: this.client?.readyAt ? true : false,
      user: this.client?.user?.tag || null,
      guilds: this.client?.guilds?.cache?.size || 0
    };
  }
}

module.exports = DiscordBot;